### PR TITLE
Add support for exchange rate rest api endpoint

### LINKED ISSUE #3579: Exchange Rate REST API
### Problem

The mirror node should expose an exchange rate REST API so users can estimate gas price or exchanges can estimate cost in USD.

### Solution

* A new exchange rate REST API `/api/v1/network/exchangerate` will be added that returns the [exchange rate](https://github.com/hashgraph/hedera-protobufs/blob/main/services/exchange_rate.proto) network file.
* Retrieve `file_data` for file ID `0.0.112`. Extract the contract bytecode [logic](https://github.com/hashgraph/hedera-mirror-node/blob/55599ea2b6a5ab460f1c1f800d1f3f506a6e2c57/hedera-mirror-rest/controllers/contractController.js#L70) to a separate fileService.
* Cache max-age should be set to 60s or more since file only updates every hour.
```json
{
  "current_rate": {
    "cent_equivalent": 596987
    "expiration_time": 1649689200
    "hbar_equivalent": 30000
  },
  "next_rate": {
    "cent_equivalent": 594920
    "expiration_time": 1649692800
    "hbar_equivalent": 30000
  },
  "timestamp": "1649689200.123456789"
}
```
Query Parameters:

- `timestamp` - Show the exchange rate as of a certain timestamp or timestamp range. Supports `eq`, `gt`, `gte`, `lt`, `lte` operators, but note that it will only return the most recent exchange rate within the given timestamp range.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Nana-EC**
**Description**:
Add Mirror Node support to retrieve the exchange rate details of the network

**Related issue(s)**:
- Add `ExchangeRateSetViewModel` and `ExchangeRateViewModel`
- Add `FileDataService` to handle  `file_data` row extraction logic
- Update `BaseService` with `buildWhereSqlStatement` and  `getFilterWhereCondition` methods currently in controller
- Add `exchangerate` to network route
- Update `networkController` with `network/exchangerate`
- Move `ContractController.fileDataQuery` to `FileDataService`

Fixes #3579 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-rest/controllers/networkController.js [hunk 1]
```diff
@@ -164,6 +124,66 @@ class NetworkController extends BaseController {
...
+          }
+
+          // to ensure most recent occurence is found convert eq to lte
+          if (utils.opsMap.eq === filter.operator) {
+            filter.operator = utils.opsMap.lte;
+          }
+
+          filterQuery.whereQuery.push(FileDataService.getFilterWhereCondition(FileData.CONSENSUS_TIMESTAMP, filter));
+          break;
+        case constants.filterKeys.ORDER:
```
Comment (steven-sheehy) [inline hedera-mirror-rest/controllers/networkController.js:155]: Why do we need to support an order parameter? It's simpler to just say we return the latest as of the given timestamp. Also, `order` isn't in the HIP.

File: hedera-mirror-rest/service/fileDataService.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    const outerWhere = innerWhere.replace('and ', `and ${FileData.tableAlias}.`);
+    return FileDataService.latestFileContentsQuery
+      .replace(FileDataService.filterInnerPlaceholder, innerWhere)
+      .replace(FileDataService.filterOuterPlaceholder, outerWhere);
+  };
+
+  getLatestFileDataContents = async (fileId, filterQueries) => {
+    const {where, params} = super.buildWhereSqlStatement(filterQueries.whereQuery, [fileId]);
+    const row = await super.getSingleRow(this.getLatestFileContentsQuery(where), params, 'getLatestFileContents');
+    return _.isNil(row) ? null : row;
```
Comment (steven-sheehy) [inline hedera-mirror-rest/service/fileDataService.js:108]: Redundant check.

```suggestion
    return row;
```

File: hedera-mirror-rest/service/fileDataService.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+class FileDataService extends BaseService {
+  static exchangeRateFileId = 112;
+
+  // placeholders to support where filtering for inner and outer calls
+  static filterInnerPlaceholder = '<filterInnerPlaceHolder>';
+  static filterOuterPlaceholder = '<filterOuterPlaceHolder>';
+
+  // retrieve the largest timestamp of the most recent create/update operation on the file
+  // using this timestamp retrieve all recent file operations and combine contents for applicable file
+  static latestFileContentsQuery = `with lastest_create as (
```
Comment (steven-sheehy) [inline hedera-mirror-rest/service/fileDataService.js:41]:
```suggestion
  static latestFileContentsQuery = `with latest_create as (
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/model/exchangeRate.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+   * Parses exchange rate into object
+   * Currently from proto, eventually from exchange_rate table
+   */
+  constructor(exchangeRate) {
+    let exchangeRateSet = {};
+
+    try {
+      exchangeRateSet = proto.ExchangeRateSet.decode(Buffer.from(exchangeRate.file_data, 'hex'));
+    } catch (error) {
+      throw new FileDecodeError(`${error.message}`);
```
Comment (xin-hedera) [inline hedera-mirror-rest/model/exchangeRate.js:37]: nit
```suggestion
      throw new FileDecodeError(error.message);
```

File: hedera-mirror-rest/errors/fileDecodeError.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    this.message = FileDecodeErrorMessage;
+    if (errorMessage !== undefined) {
+      this.message += errorMessage === undefined ? '' : `. Error: '${errorMessage}'`;
+    }
+  }
+}
+
+module.exports = {
+  FileDecodeError,
+};
```
Comment (xin-hedera) [inline hedera-mirror-rest/errors/fileDecodeError.js:38]: we should have done this for other error classes too.

```suggestion
module.exports = FileDecodeError;
```

File: hedera-mirror-rest/controllers/networkController.js [hunk 6]
```diff
@@ -194,6 +214,46 @@ class NetworkController extends BaseController {
...
+      const {rows} = result;
+
+      if (rows.length !== 1 || !rows[0].consensus_timestamp) {
+        throw new NotFoundError('Not found');
+      }
+
+      res.locals[constants.responseDataLabel] = new NetworkSupplyViewModel(rows[0], NetworkController.totalSupply);
+      logger.debug(`getSupply returning ${result.rows.length} entries`);
+    });
+  };
```
Comment (xin-hedera) [inline hedera-mirror-rest/controllers/networkController.js:256]: nit, use `async/await` style

File: hedera-mirror-rest/controllers/networkController.js [hunk 7]
```diff
@@ -164,6 +124,66 @@ class NetworkController extends BaseController {
...
+
+    const filterQueries = this.extractExchangeRateQuery(filters);
+
+    const exchangeRate = await FileDataService.getExchangeRate(filterQueries);
+
+    if (_.isNil(exchangeRate)) {
+      throw new NotFoundError('Not found');
+    }
+
+    res.send(new ExchangeRateSetViewModel(exchangeRate));
```
Comment (xin-hedera) [inline hedera-mirror-rest/controllers/networkController.js:184]: should set `res.locals[constants.responseDataLabel]` instead

File: hedera-mirror-rest/errors/fileDecodeError.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+const FileDecodeErrorMessage =
+  'Failed to decode file contents. Ensure timestamp filters cover the complete file create/update and append transactions';
+
+class FileDecodeError extends Error {
+  constructor(errorMessage) {
+    super();
+    this.message = FileDecodeErrorMessage;
+    if (errorMessage !== undefined) {
+      this.message += errorMessage === undefined ? '' : `. Error: '${errorMessage}'`;
+    }
```
Comment (xin-hedera) [inline hedera-mirror-rest/errors/fileDecodeError.js:32]:
```suggestion
    if (errorMessage !== undefined) {
      this.message += `. Error: '${errorMessage}'`;
    }
```

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/controllers/networkController.js (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js:256]: Not new code, just reordered so I haven't touched any of the logic here

File: hedera-mirror-rest/controllers/networkController.js [hunk 9]
```diff
@@ -194,6 +214,46 @@ class NetworkController extends BaseController {
 
     res.locals[constants.responseDataLabel] = response;
   };
+
+  /**
```
Comment (Nana-EC) [inline hedera-mirror-rest/controllers/networkController.js:218]: Just moved this function from top of the file down

File: hedera-mirror-rest/controllers/networkController.js (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js:184]: Hehe i added that logic too. Was testing out why I had that in place of a `res.send` which one of the current PRs out had and forgot to change back

File: hedera-mirror-rest/controllers/networkController.js (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js:155]: Will remove.

File: hedera-mirror-rest/errors/fileDecodeError.js (hunk 8 above)
Comment (Nana-EC) [reply hedera-mirror-rest/errors/fileDecodeError.js]: Updated

File: hedera-mirror-rest/controllers/networkController.js (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js]: Fixed

File: hedera-mirror-rest/controllers/networkController.js (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js]: Updated as a plus.
This method logic should be updated to call a `NetworkNodeServoce.getSupply()`

File: hedera-mirror-rest/errors/fileDecodeError.js (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-rest/errors/fileDecodeError.js]: Updated

File: hedera-mirror-rest/model/exchangeRate.js (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-rest/model/exchangeRate.js]: Updated

File: hedera-mirror-rest/service/fileDataService.js (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-rest/service/fileDataService.js]: Updated

File: hedera-mirror-rest/service/fileDataService.js (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-rest/service/fileDataService.js]: Updated

File: hedera-mirror-rest/controllers/networkController.js (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js]: Removed

File: hedera-mirror-rest/controllers/networkController.js [hunk 10]
```diff
@@ -164,6 +124,59 @@ class NetworkController extends BaseController {
     };
   };
 
+  extractExchangeRateQuery = (filters) => {
+    // get latest rate only
+    const filterQuery = {
+      order: constants.orderFilterValues.DESC,
+      limit: 1,
```
Comment (xin-hedera) [inline hedera-mirror-rest/controllers/networkController.js:131]: nit `order` / `limit` not needed at all

**Review by xin-hedera — DISMISSED** [review]
LGTM. just one nit

**Review by steven-sheehy — DISMISSED** [review]
LGTM

File: hedera-mirror-rest/controllers/networkController.js (hunk 10 above)
Comment (xin-hedera) [reply hedera-mirror-rest/controllers/networkController.js:131]: I'd say we should remove it from this function no matter what. In addition, we should move away from building SQL expressions / clauses in controllers.

Comment (steven-sheehy) [reply hedera-mirror-rest/controllers/networkController.js]: Agreed. It's an anti-pattern to have any SQL in the controller.

**Review by xin-hedera — APPROVED** [review]
LGTM

File: hedera-mirror-rest/controllers/networkController.js (hunk 10 above)
Comment (Nana-EC) [reply hedera-mirror-rest/controllers/networkController.js]: Removed, btw

**Pull request merged**